presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometrySerde.java

```diff
@@ -32,319 +32,206 @@
+        }
 
-        output.writeInt(numPoints);
-        for (Coordinate coordinate : geometry.getCoordinates()) {
-            writeCoordinate(coordinate, output);
+        if (multitype) {
+            return GEOMETRY_FACTORY.createMultiLineString(lineStrings);
+        }
+        else {
+            verify(lineStrings.length == 1);
+            return lineStrings[0];
         }
     }
 
-    private static Geometry readPolyline(SliceInput input)
+    private static Geometry readPolygon(SliceInput input, boolean multitype)
     {
-        requireNonNull(input, "input is null");
+        skipEsriType(input);
         skipEnvelope(input);
         int partCount = input.readInt();
         if (partCount == 0) {
-            return GEOMETRY_FACTORY.createLineString();
+            if (multitype) {
+                return GEOMETRY_FACTORY.createPolygon();
+            }
+            else {
+                return GEOMETRY_FACTORY.createPolygon();
+            }
         }
 
         int pointCount = input.readInt();
-        LineString[] lineStrings = new LineString[partCount];
+        int[] startIndexes = new int[partCount];
+        for (int i = 0; i < partCount; i++) {
+            startIndexes[i] = input.readInt();
+        }
 
         int[] partLengths = new int[partCount];
-        input.readInt();    // skip first index; it is always 0
         if (partCount > 1) {
-            partLengths[0] = input.readInt();
+            partLengths[0] = startIndexes[1];
             for (int i = 1; i < partCount - 1; i++) {
-                partLengths[i] = input.readInt() - partLengths[i - 1];
+                partLengths[i] = startIndexes[i + 1] - startIndexes[i];
             }
-            partLengths[partCount - 1] = pointCount - partLengths[partCount - 2];
-        }
-        else {
-            partLengths[0] = pointCount;
         }
+        partLengths[partCount - 1] = pointCount - startIndexes[partCount - 1];
 
+        LinearRing shell = null;
+        List<LinearRing> holes = new ArrayList<>();
+        List<Polygon> polygons = new ArrayList<>();
         for (int i = 0; i < partCount; i++) {
-            lineStrings[i] = GEOMETRY_FACTORY.createLineString(readCoordinates(input, partLengths[i]));
+            Coordinate[] coordinates = readCoordinates(input, partLengths[i]);
+            if (isClockwise(coordinates)) {
+                // next polygon has started
+                if (shell != null) {
+                    polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
+                    holes.clear();
+                }
+                else {
+                    verify(holes.isEmpty(), "shell is null but holes found");
+                }
+                shell = GEOMETRY_FACTORY.createLinearRing(coordinates);
+            }
+            else {
+                verify(shell != null, "shell is null but hole found");
```

Comment:
Is it possible to enter this `else` branch in the first iteration (`isClockwise(coordinates)` is false)? If yes, then this assertion will fail.

Reply:
Yes, it is possible.

Shells are stored clockwise. Holes are stored counterclockwise. Shell is stored first followed by the holes. Whenever we find a shell - that means that a new polygon has started. However if the very first shape is a hole - that means that serialized geometry is corrupted.